src/allmydata/test/test_introducer.py

```diff
@@ -314,6 +316,64 @@ def test_duplicate(self):
 
 NICKNAME = u"n\u00EDickname-%s" # LATIN SMALL LETTER I WITH ACUTE
 
+def foolscapEndpointForPortNumber(portnum):
+    """
+    Create an endpoint that can be passed to ``Tub.listen``.
+
+    :param portnum: Either an integer port number indicating which TCP/IPv4
+        port number the endpoint should bind or ``None`` to automatically
+        allocate such a port number.
+
+    :return: A two-tuple of the integer port number allocated and a
+        Foolscap-compatible endpoint object.
+    """
+    if portnum is None:
+        from twisted.internet import reactor
+        from twisted.internet.interfaces import IReactorSocket
+        if IReactorSocket.providedBy(reactor):
+            # On POSIX we can take this very safe approach of binding the
+            # actual socket to an address.  Once the bind succeeds here, we're
+            # no longer subject to any future EADDRINUSE problems.
+            import fcntl
+            s = socket()
+            try:
+                s.bind(('', 0))
+                portnum = s.getsockname()[1]
+                s.listen(1)
+                fd = os.dup(s.fileno())
+                flags = fcntl.fcntl(fd, fcntl.F_GETFD)
+                flags = flags | os.O_NONBLOCK | fcntl.FD_CLOEXEC
+                fcntl.fcntl(fd, fcntl.F_SETFD, flags)
+                return (
+                    portnum,
+                    AdoptedStreamServerEndpoint(reactor, fd, AF_INET),
+                )
+            finally:
+                s.close()
+        else:
+            # Get a random port number and fall through.  This is necessary on
+            # Windows where Twisted doesn't offer IReactorSocket.  This
+            # approach is error prone for the reasons described on
```

Comment:
Ah, my previous comment's question is answered :) -- so maybe a comment that the IReactorSocket check is also telling us if we're POSIX vs Windows?

Reply:
Yea ... for now, anyway.  I suppose if someone eventually implements IReactorSocket for some reactor on Windows then this might no longer be the case.  And, indeed, I believe `fcntl` is not available on Windows (https://docs.python.org/2/library/index.html puts it in the "Unix Specific Services" section which I take as a hint in this direction as well).

Reply:
I am reminded of `twisted.python.reflect.requireModule`, though, which supports an idiom I prefer for maybe-not-available modules.  So I'll do that.